blivet/mounts.py

```diff
@@ -0,0 +1,137 @@
+    def add(self, devspec, subvolspec=None):
+        """ Add device to cache
+
+            :param devscpec: device specification, eg. "/dev/vda1"
+            :type devspec: str
+            :param subvolspec: btrfs subvolume specification, eg. ID or name
+            :type subvolspec: str
+
+        """
+
+        self.mountpoints[(devspec, subvolspec)] = None
+        self.newDevice = True
+
+    def remove(self, devspec, subvolspec=None):
+        """ Remove device from cache
+
+            :param devscpec: device specification, eg. "/dev/vda1"
+            :type devspec: str
+            :param subvolspec: btrfs subvolume specification, eg. ID or name
+            :type subvolspec: str
+
+        """
+
+        if (devspec, subvolspec) in self.mountpoints:
+            del self.mountpoints[(devspec, subvolspec)]
+
+    def clear(self):
+        """ Clear cache
+        """
+
+        for key in self.mountpoints.keys():
+            self.mountpoints[key] = None
+
+        self._getActiveMounts()
+
+    def getMountpoint(self, devspec, subvolspec=None):
+        """ Get mountpoint for selected device
+
+            :param devscpec: device specification, eg. "/dev/vda1"
+            :type devspec: str
+            :param subvolspec: btrfs subvolume specification, eg. ID or name
+            :type subvolspec: str
+            :returns: mountpoint (path)
+            :rtype: str
+
+        """
+
+        self._cacheCheck()
+
+        if (devspec, subvolspec) not in self.mountpoints.keys():
+            return None
+
+        else:
+            return self.mountpoints[(devspec, subvolspec)]
+
+    def _getActiveMounts(self):
+
+        for line in open("/proc/mounts").readlines():
+            try:
+                (devspec, mountpoint, fstype, options, _rest) = line.split(None, 4)
+            except ValueError:
+                log.error("failed to parse /proc/mounts line: %s", line)
+                continue
+
+            if fstype == "btrfs":
+                # get the subvol name from /proc/self/mountinfo
+                for line in open("/proc/self/mountinfo").readlines():
+                    fields = line.split()
+                    _subvol = fields[3]
+                    _mountpoint = fields[4]
+                    _devspec = fields[9]
+                    if _mountpoint == mountpoint and _devspec == devspec:
+                        # empty _subvol[1:] means it is a top-level volume
+                        subvolspec = _subvol[1:] or 5
+
+                        fmt = self._resolveFormat(devspec, subvolspec)
+
+            else:
+                fmt = self._resolveFormat(devspec)
+
+            if fmt:
+                self.mountpoints[fmt] = mountpoint
+
+    def _resolveFormat(self, devspec, subvolspec=None):
+
+        for fmt in self.mountpoints.keys():
+            if fmt[0] == devspec:
+                if not fmt[1]:
+                    return fmt
+
+                elif fmt[1] == subvolspec:
+                    return fmt
```

Comment:
```
return next((f for (d, s) in self.mountpoints.keys() if d == devspec and (not s or s == subvolspec)), None)
```
should do the same thing and also explicitly returns None.

This method is probably worth a doc string.